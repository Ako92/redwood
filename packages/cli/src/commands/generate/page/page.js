import { execSync } from 'child_process'

import Listr from 'listr'
import terminalLink from 'terminal-link'
import { getProject } from '@redwoodjs/structure'
import paramCase from 'param-case'
import { writeFilesTask } from 'src/x/listr/writeFilesTask'

import c from 'src/lib/colors'

import { templateForComponentFile } from '../helpers'

const COMPONENT_SUFFIX = 'Page'
const REDWOOD_WEB_PATH_NAME = 'pages'

export const templates = (
  { name, path, routes },
  { project = getProject() } = {}
) => {
  const page = templateForComponentFile(
    {
      name,
      suffix: COMPONENT_SUFFIX,
      webPathSection: REDWOOD_WEB_PATH_NAME,
      generator: 'page',
      templatePath: 'page.js.template',
      templateVars: { path },
    },
    project.pathHelper
  )
  const test = templateForComponentFile(
    {
      name,
      suffix: COMPONENT_SUFFIX,
      extension: '.test.js',
      webPathSection: REDWOOD_WEB_PATH_NAME,
      generator: 'page',
      templatePath: 'test.js.template',
      templateVars: { path },
    },
    project.pathHelper
  )
  const stories = templateForComponentFile(
    {
      name,
      suffix: COMPONENT_SUFFIX,
      extension: '.stories.js',
      webPathSection: REDWOOD_WEB_PATH_NAME,
      generator: 'page',
      templatePath: 'stories.js.template',
      templateVars: { path },
    },
    project.pathHelper
  )

  return [
    {
      path: stories[0],
      contents: stories[1],
    },
    {
      path: test[0],
      contents: test[1],
    },
    {
      path: page[0],
      contents: page[1],
    },
    {
      path: project.pathHelper.web.routes,
      contents: project.router.createRouterString(routes),
      overwrite: true, // file exists
      doNotDelete: true, // do not delete
    },
  ]
}

export const command = 'page <name> [path]'
export const description = 'Generate a page and route component'
export const builder = (yargs) => {
  yargs
    .positional('name', {
      description: 'Name of the page',
      type: 'string',
    })
    .positional('path', {
      description: 'URL path to the page. Defaults to name',
      type: 'string',
    })
    .option('force', {
      alias: 'f',
      default: false,
      description: 'Overwrite existing files',
      type: 'boolean',
    })
    .option('json', {
      default: false,
      description: 'Output as JSON',
      type: 'boolean',
    })
    .epilogue(
      `Also see the ${terminalLink(
        'Redwood CLI Reference',
        'https://redwoodjs.com/reference/command-line-interface#generate-page'
      )}`
    )
}

const cleanPathArgForWindows = (path) => {
  if (path && process.platform === 'win32') {
    // running `yarn rw g page home /` on Windows using GitBash
    // POSIX-to-Windows path conversion will kick in.
    // See https://github.com/git-for-windows/build-extra/blob/d715c9e/ReleaseNotes.md
    // As a workaround we try to detect when this has happened, and reverse
    // the action
    try {
      // `cygpath -m /` will return something like 'C:/Program Files/Git/\n'
      const slashPath = execSync('cygpath -m /', {
        stdio: ['ignore', 'pipe', 'ignore'],
      })
        .toString()
        .trim()

      // `yarn rw g page home /` =>
      //   page === 'C:/Program Files/Git'
      // `yarn rw g page about /about` =>
      //   page === 'C:/Program Files/Git/about'
      // Sometimes there is a / after 'Git' to match, sometimes there isn't
      return path.replace(new RegExp(`^${slashPath}?`), '/')
    } catch {
      // probably using PowerShell or cmd, in which case no special handling
      // is needed
    }
  }
}

export const handler = async ({ name, path, force, json }) => {
  path = cleanPathArgForWindows(path)
  // Generate a path if the user didn't supply one
  path = path ?? `/${paramCase(name)}`

  const newRoute = getProject().router.createRouteString(name, path)
  const files = await templates({ name, path, routes: [newRoute] })

  if (json) {
    console.log(JSON.stringify(files, 2))
  } else {
    const tasks = new Listr(
      [
        {
          title: 'Generating page files...',
          task: async () => {
            return writeFilesTask(files, { overwrite: force })
          },
        },
      ].filter(Boolean),
      { collapse: false }
    )

    try {
      await tasks.run()
    } catch (e) {
      console.log(c.error(e.message))
    }
  }
}
