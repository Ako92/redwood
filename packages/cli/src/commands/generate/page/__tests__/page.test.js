import path from 'path'

import { getProject } from '@redwoodjs/structure'

import { templates } from '../page'

describe('page generator', () => {
  const fx = path.resolve(
    __dirname,
    '../../../../../../../__fixtures__/example-todo-main'
  )
  const project = getProject(fx)

  it('creates the correct templates', () => {
    const name = 'Home'
    const path = '/home'
    const route = project.router.createRouteString(name, path)
    const newTemplates = templates({ name, path, routes: [route] }, { project })
    // Test the path output.
    expect(newTemplates.map((t) => t.path.replace(fx, '')))
      .toMatchInlineSnapshot(`
    Array [
      "/web/src/pages/HomePage/HomePage.stories.js",
      "/web/src/pages/HomePage/HomePage.test.js",
      "/web/src/pages/HomePage/HomePage.js",
      "/web/src/Routes.js",
    ]
  `)

    // write a snapshot for each generated template.
    for (const t of newTemplates) {
      expect(t.contents).toMatchSnapshot(t.path.replace(fx, ''))
    }
  })
})
