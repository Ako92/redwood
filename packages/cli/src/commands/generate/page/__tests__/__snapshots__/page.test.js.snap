// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page generator creates the correct templates: /web/src/Routes.js 1`] = `
"// In this file, all Page components from 'src/pages\` are auto-imported. Nested
// directories are supported, and should be uppercase. Each subdirectory will be
// prepended onto the component name.
//
// Examples:
//
// 'src/pages/HomePage/HomePage.js'         -> HomePage
// 'src/pages/Admin/BooksPage/BooksPage.js' -> AdminBooksPage

import { Router, Route } from '@redwoodjs/router'

const Routes = () => {
  return (
    <Router>
      <Route path=\\"/home\\" page={HomePage} name=\\"home\\" />
      <Route path=\\"/\\" page={HomePage} name=\\"home\\" />
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;

exports[`page generator creates the correct templates: /web/src/pages/HomePage/HomePage.js 1`] = `
"import { Link } from \\"@redwoodjs/router\\";

const HomePage = () => {
  return (
    <>
      <h1>HomePage</h1>
      <p>Find me in \\"./web/src/pages/HomePage/HomePage.js\\"</p>
      <p>
        My default route is named \\"home\\", link to me with \`
        <Link to=\\"home\\">routes.home()</Link>\`
      </p>
    </>
  );
};

export default HomePage;
"
`;

exports[`page generator creates the correct templates: /web/src/pages/HomePage/HomePage.stories.js 1`] = `
"import HomePage from \\"./HomePage\\";

export const generated = () => {
  return <HomePage />;
};

export default { title: \\"Pages/HomePage\\" };
"
`;

exports[`page generator creates the correct templates: /web/src/pages/HomePage/HomePage.test.js 1`] = `
"import { render } from \\"@redwoodjs/testing\\";

import HomePage from \\"./HomePage\\";

describe(\\"HomePage\\", () => {
  it(\\"renders successfully\\", () => {
    expect(() => {
      render(<HomePage />);
    }).not.toThrow();
  });
});
"
`;
